Improve Enhancements Role Handbook with test freeze, PRR process, and… (#2894)
kubernetes/sig-release: release-team/role-handbooks/enhancements/README.md

  - [Slack](#slack)
- [Process](#process)
  - [Standards](#standards)
  - [Enhancement KEP Status](#enhancement-kep-status)
  - [PRR Reviews](#prr-reviews)

>     - [PRR Freeze](#prr-freeze)

Reviewer: can you check the order here? they should be swapped to match the order below

```suggestion
- [PRR Reviews](#prr-reviews)
    - [PRR Freeze](#prr-freeze)
- [Enhancement KEP Status](#enhancement-kep-status)
```

- [Process](#process)
  - [Standards](#standards)
  - [PRR Reviews](#prr-reviews)
    - [PRR Freeze](#prr-freeze)
    - [PRR Freeze Review Process](#prr-freeze-review-process)

>     - [Enhancement KEP Status](#enhancement-kep-status)

Reviewer: just checking, did you mean to indent `Enhancement KEP Status` under `PRR Reviews`? or should it be a level higher?

  approver: @<gh-handle-of-PRR-approver>
```

#### PRR Freeze

> See [PRR Freeze](https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#prr-freeze) for the definition and deadline details.

Reviewer: ```suggestion
See [PRR Freeze](/releases/release_phases.md#prr-freeze) for the definition and deadline details.
```
let's use the local path for all the sig-release links

> #### Enhancement KEP Status

Reviewer: ```suggestion
### Enhancement KEP Status
```

  approver: @<gh-handle-of-PRR-approver>
```

> #### PRR Freeze

Reviewer: the content in this section is also mentioned in https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#prr-freeze which we link to below, so maybe we don't need it here but up to you if you want to keep a copy here in the enhancements handbook. i'm a little worried it would get out of date if `release_phases.md` changes though

The PRR freeze is a hard deadline happening one week before the [Enhancements Freeze](https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#enhancements-freeze). KEPs that will miss this deadline require an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) or they will be removed from the milestone.

As described by the PRR team [here](https://groups.google.com/a/kubernetes.io/g/dev/c/CQ33yPqp-H4/m/hHO-NaQiAQAJ):

- By the PRR freeze date, KEP authors must have completed the PRR questionnaire for all opted-in enhancements (KEPs).
- It is important to note that the PRR freeze **_does not_** mean that the KEPs need to have received an approval or even a review from the PRR team by this date.
- To emphasize, the sole requirement is that all opted-in KEPs have their PRR questionnaires answered by the deadline, to ensure the PRR team has sufficient time to review them by Enhancements Freeze.

> > Note: KEPs that doesn't complete the PRR questionnaire and get back to the reviews from the PRR team are removed from the milestone during the enhancements freeze party. The PRR freeze deadline exists as a reminder for this.

Reviewer: ```suggestion
> Note: KEPs that don't complete the PRR questionnaire and respond to the reviews from the PRR team are removed from the milestone during the PRR freeze party. The PRR freeze deadline exists as a reminder for this.
```

> #### PRR Freeze Review Process

Reviewer: this is redundant with the requirements listed in the Milestone Activities timeline [below](https://github.com/kubernetes/sig-release/tree/master/release-team/role-handbooks/enhancements#before-prr-freeze) so maybe we don't need to add this section here too. in my open PR [here](https://github.com/jenshu/sig-release/blob/93f80aeeb305eb683913f03d66df3a893f440f1d/release-team/role-handbooks/enhancements/README.md#prr-freeze-tasks) i have also refactored it a bit to clean it up

- Provide updates during release team meetings
  - Use the `KEPs by Stage` insight from the Enhancement Tracking Board ([example](https://github.com/orgs/kubernetes/projects/98/insights/3))
- Notify folks in [#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) about KEPs waiting for PRR reviews with the project board filtered for KEPs missing PRR assignees.
  
  **How to notify #prod-readiness about KEPs waiting for PRR reviews:**
  
  1. **Filter the project board** to show KEPs missing PRR assignees:
     - Go to the Enhancements Tracking Board
     - Apply filters: `PRR Status` = "Not Started" or "In Progress"
     - This will show KEPs that need PRR attention
     

>   2. **Timing**: Send notifications typically 1-2 weeks before enhancement freeze to give PRR team sufficient time to review

Reviewer 2: ```suggestion
  2. **Timing**: Send notifications typically 1-2 weeks before PRR freeze to give PRR team sufficient time to review
```

- Since some KEPs may have been removed from the milestone after PRR Freeze, re-balance the issue assignments (`Enhancements Contact`) across the team as needed.
- For each issue that is already `Tracked for PRR Freeze`:
  - Evaluate if the enhancement satisfies all the [Enhancements Freeze requirements](/releases/release_phases.md#enhancements-freeze) for inclusion in the current release.

>     - Comment on the Issue with a status updating using one of the [Enhancements Freeze communication templates](./communication-templates.md#enhancement-freeze-templates)

Reviewer: typo - should be `#enhancements-freeze-templates`

    - Update the `Status` field for this Enhancement in the Enhancement Tracking Board (`Tracked for enhancements freeze` or `At risk for enhancements freeze`)
    - KEPs targeting `stable` will need to be marked as `implemented` after code PRs are merged. This will need to be verified after the code freeze.
    - *Note*: Refer to the [README template](https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/README.md?plain=1) to determine which sections are required based on the stage (`alpha`/`beta`/`stable`) that the enhancement is targeting. It is not the Enhancement team's responsibility to validate the correctness/completeness of the README content; we only need to make sure that each required section is present.

> - Reach out to each [SIG on Slack](https://github.com/kubernetes/community/blob/master/sig-list.md) tagging one or more chairs or technical leads to communicate the Enhancements statuses. See the example [SIG Outreach communication templates](./communication-templates.md#sig-outreach-templates-optional).

Reviewer: this can be removed, as it's covered here already: https://github.com/kubernetes/sig-release/tree/master/release-team/role-handbooks/enhancements#sig-outreach

- **Ensure the Code Freeze PR is Ready to Merge**:
  The Code Freeze PR should be ready for merge (i.e., it has both `lgtm` and `approved` labels) **on or before the date of Code Freeze**.
  **Recommendation:** Work with the Release Branch Manager to get the PR created ahead of time and ensure all **required labels** (`lgtm`, `approved`) are in place. They can add a `/hold` to prevent the PR from merging early - then simply lift the hold in the Code Freeze Party.

- **Reference Code Freeze PRs**: Below are examples of PRs used to enforce Code Freeze in previous releases:
  - [PR Enforcing Code Freeze - v1.33](https://github.com/kubernetes/test-infra/pull/34532)
  - [PR Enforcing Code Freeze - v1.32](https://github.com/kubernetes/test-infra/pull/33778)
  - [PR Enforcing Code Freeze - v1.31](https://github.com/kubernetes/test-infra/pull/33049)

Read https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#code-freeze for more information on the Code Freeze.

#### Understanding Test Freeze

> Test freeze occurs simultaneously with [Code Freeze](https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#code-freeze). For the Enhancements team, this means:

Reviewer: ```suggestion
[Test Freeze](/releases/release_phases.md#test-freeze) occurs simultaneously with [Code Freeze](/releases/release_phases.md#code-freeze). For the Enhancements team, this means:
```

#### Succession

> **Important**: All handbook updates or changes must be merged or addressed before the new release cycle begins.

- Select who will be the new Enhancements Lead for the next release. Shadows should be the first source pool. If none are available to lead then look externally through other release team members or members of [SIG Architecture Enhancements Subproject](https://github.com/kubernetes/community/blob/master/sig-architecture/README.md#enhancements)
- Make sure to get the selection approved by the Release Team Lead and Subproject Leads before giving final confirmation to the candidate.
- Before the next release starts, do a handoff call with the incoming Enhancements Lead. Walk through all the tasks that the Enhancements Lead needs to do throughout the cycle, and provide any other information/advice that would help prepare them to lead the team successfully!
- Continually work to improve Enhancements process

> - Review / update documentation as the release cycle ends

Reviewer: i think these are already covered [below](https://github.com/kubernetes/sig-release/tree/master/release-team/role-handbooks/enhancements#process-and-documentation-improvements), so can be removed

- Close issues marked as stable that made it into the release, only after the corresponding KEPs have been marked `Implemented`
- Close milestones that are complete

> - Cleanup old milestones

Reviewer: already covered [here](https://github.com/kubernetes/sig-release/tree/master/release-team/role-handbooks/enhancements#tracking-board-cleanup), can be removed

> - Open a pull request to add the `exceptions.yaml` file with all the KEPs which applied for an exception during enhancements freeze and code freeze

Reviewer: can remove this as we don't use exceptions.yaml anymore; we should open a separate PR to add the new exception tracking process
